### ISSUE TITLE
Consistent treatment of virtualNode

### ISSUE THREAD

**Issue by dylanb** [dequelabs/axe-core#435]
Some utils (shorthand for utils and commons) are using the virtualNode whereas some are using node and then looking up the virtualNode in the flattened tree cache.

This is inconsistent but also raises the following concerns:

1. Performance of lookups
1. Reliance on cache making utilities less usable for other purposes
1. Memory leaks if cache left uncleared
1. No transparency when DOM can change and cache is out of date
1. Backwards compatibility of existing rule sets or custom rules

I propose that we do the following:

1. Create new utils that have a unique name that identifies them as needing a virtualNode
1. Create simple utils with the current names that perform the virtualNode lookup and then call the new util (for backwards compatibility)

In the builtin rules that will ship with this version of axe-core:

1. Use the new utils that require the virtualNode in every context where the virtualNode is available (i.e. every check and matches function) and
1. Call the lookup util in all other cases

@marcysutton @WilcoFiers thoughts?

**Comment by marcysutton** [comment]
I like the idea of bridging contexts with new utils to smooth out use of `virtualNode`, especially for backwards compatibility. I'm of the opinion that we should go for it. Is there any downside?

For memory leaks, I wonder if we could implement tooling to help detect them? I found a few Node utilities that might make this doable. Here's one: https://www.npmjs.com/package/leakage

**Comment by dylanb** [comment]
Well I think what I am suggesting is to only support backwards compatibility in the context of rules/checks and not in any other context - thereby solving that problem by throwing an exception and forcing the caller to understand what is going on.

**Comment by WilcoFiers** [comment]
I'm not entirely clear on what you are suggesting. Can you elaborate on this a bit @dylanb?
```
1. Create new utils that have a unique name that identifies them as needing a virtualNode
2. Create simple utils with the current names that perform the virtualNode lookup and then call the new util (for backwards compatibility)
```

**Comment by dylanb** [comment]
For example: `axe.commons.text.accessibleText()`

```
text.accessibleText = function (element, inLabelledbyContext) {
  let virtualNode = axe.utils.getNodeFromTree(axe._tree[0]);
  return text.virtualAccessibleText(virtualNode, inLabelledbyContext);
}

text.virtualAccessibleText = function (element, inLabelledbyContext) {
  // rest of the code as in your PR is in here minus the lookup
}
```

**Comment by marcysutton** [comment]
Closed with #455 and #471.
